vyper/venom/analysis/available_expression.py

```diff
@@ -0,0 +1,272 @@
+from collections import deque
+from dataclasses import dataclass
+from functools import cached_property
+
+from vyper.utils import OrderedSet
+from vyper.venom.analysis.analysis import IRAnalysesCache, IRAnalysis
+from vyper.venom.analysis.cfg import CFGAnalysis
+from vyper.venom.analysis.dfg import DFGAnalysis
+from vyper.venom.basicblock import (
+    BB_TERMINATORS,
+    IRBasicBlock,
+    IRInstruction,
+    IROperand,
+    IRVariable,
+)
+from vyper.venom.context import IRFunction
+from vyper.venom.effects import EMPTY, Effects
+
+_MAX_DEPTH = 5
+_MIN_DEPTH = 2
+
+UNINTERESTING_OPCODES = ["store", "param", "offset", "phi", "nop"]
+_IDEMPOTENT_INSTRUCTIONS = ["log", "call", "staticcall", "delegatecall", "invoke"]
+
+
+@dataclass
+class _Expression:
+    first_inst: IRInstruction
+    opcode: str
+    operands: list["IROperand | _Expression"]
+    ignore_msize: bool
+
+    # equality for lattices only based on first_inst
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, _Expression):
+            return False
+
+        return self.first_inst == other.first_inst
+
+    # Full equality for expressions based on opcode and operands
+    def same(self, other) -> bool:
+        if type(self) is not type(other):
+            return False
+
+        if self.opcode != other.opcode:
+            return False
+
+        # Early return special case for commutative instructions
+        if self.is_commutative:
+            if self.operands[0].same(other.operands[1]) and self.operands[1].same(
+                other.operands[0]
+            ):
+                return True
+
+        # General case
+        for self_op, other_op in zip(self.operands, other.operands):
+            if not self_op.same(other_op):
+                return False
```

Comment:
i think this is a performance problem. i think the way to get around it is to map all expressions to ids. then we can construct equivalence sets in O(n) instead of O(n^2) time.

Reply:
i think this approach will make the code cleaner too